OrbitQt/GgpInstanceItemModel.cpp

```diff
@@ -0,0 +1,153 @@
+    : QAbstractItemModel(parent), instances_(std::move(instances)) {
+  std::sort(instances_.begin(), instances_.end(), &GgpInstance::CmpById);
+}
+
+int GgpInstanceItemModel::columnCount(const QModelIndex& parent) const {
+  return parent.isValid() ? 0 : static_cast<int>(Columns::NumberOfColumns);
+}
+
+QVariant GgpInstanceItemModel::data(const QModelIndex& index, int role) const {
+  CHECK(index.isValid());
+  CHECK(index.model() == this);
+  CHECK(index.row() < instances_.size());  // instances_.size());
+
+  const GgpInstance& current_instance = instances_[index.row()];
+
+  if (role == Qt::UserRole) return QVariant::fromValue(current_instance);
+
+  if (role != Qt::DisplayRole) return {};
+
+  switch (static_cast<Columns>(index.column())) {
+    case Columns::DisplayName:
+      return current_instance.display_name;
+    case Columns::ID:
+      return current_instance.id;
+    case Columns::IPAddress:
+      return current_instance.ip_address;
+    case Columns::LastUpdated:
+      return current_instance.last_updated.toString(Qt::TextDate);
+    case Columns::Owner:
+      return current_instance.owner;
+    case Columns::Pool:
+      return current_instance.pool;
+    case Columns::NumberOfColumns:
+      CHECK(false);
+      return {};
+  }
+
+  CHECK(false);  // That means, someone (me?) forgot a column.
+  return {};
+}
+
+QModelIndex GgpInstanceItemModel::index(int row, int col,
+                                        const QModelIndex& parent) const {
+  if (parent.isValid()) return {};
+  if (row < 0 || row >= instances_.size()) return {};
+  if (col < 0 || col >= static_cast<int>(Columns::NumberOfColumns)) return {};
+
+  return createIndex(row, col, nullptr);
+}
+
+QVariant GgpInstanceItemModel::headerData(int section,
+                                          Qt::Orientation orientation,
+                                          int role) const {
+  if (role != Qt::DisplayRole) return {};
+  if (orientation != Qt::Horizontal) return {};
+  if (section < 0 || section >= static_cast<int>(Columns::NumberOfColumns)) {
+    return {};
+  }
+
+  switch (static_cast<Columns>(section)) {
+    case Columns::DisplayName:
+      return QLatin1String("Display Name");
+    case Columns::ID:
+      return QLatin1String("ID");
+    case Columns::IPAddress:
+      return QLatin1String("IP Address");
+    case Columns::LastUpdated:
+      return QLatin1String("Last Updated");
+    case Columns::Owner:
+      return QLatin1String("Owner");
+    case Columns::Pool:
+      return QLatin1String("Pool");
+    case Columns::NumberOfColumns:
+      CHECK(false);
+      return {};
+  }
+
+  CHECK(false);
+  return {};
+}
+
+QModelIndex GgpInstanceItemModel::parent(const QModelIndex& /*child*/) const {
+  return {};
+}
+
+int GgpInstanceItemModel::rowCount(const QModelIndex& parent) const {
+  return parent.isValid() ? 0 : instances_.size();
+}
+
+void GgpInstanceItemModel::SetInstances(QVector<GgpInstance> new_instances) {
```

Comment:
This method seems overly complicated.  It seems new_instances can contain instances that were there previously, could we simply wipe the old ones and replace by the new ones?

Reply:
Maybe I can explain that.

This method is implementing the recommended way to update a Qt item model, when the data is coming from an external source.

Just deleting everything and adding all items again might seem to work but is not preserving semantics and that leads to the following problems:
* Selections are not preserved
* Scroll positions are not preserved
* Ongoing edits are discarded
* Accessibility software cannot recognize the actual change.

Most of these problems are not really important for our use case. Nevertheless this scheme is good and common practice and I believe we should keep it.

Reply:
Ok, sounds good then.